### PR TITLE
Fix MongoClient handling in Mongo Panache

### LINKED ISSUE #10812: Can't use MongoDB with Panache with multiple clients on the same instance
**Describe the bug**
I have a MongoDB instance and several databases.
I don't manage to use MongoDB with Panache for several entities on different databases.

Here a simplified version of my Kotlin code :

```kotlin
@MongoEntity(clientName = "db1", collection = "c1")
data class C1 @BsonCreator constructor(
  @BsonId var id: ObjectId = ObjectId()
) : PanacheMongoEntityBase()
```

My application.properties:
```properties
quarkus.mongodb.db1.connection-string=mongodb://127.0.0.1:27017
quarkus.mongodb.db1.database=db1-name
```
I have the following error:
```
org.jboss.resteasy.spi.UnhandledException: java.util.NoSuchElementException: Property quarkus.mongodb.database not found
```
I was expected that the database name was retrieved from quarkus.mongodb.db1.database.
Am I missing something?

**To Reproduce**
Here is a [reproducer](https://github.com/ekans/quarkus-mongodb-panache-reproducer)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:
```
Linux Envy 5.4.0-40-generic #44-Ubuntu SMP Tue Jun 23 00:01:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
 - Output of `java -version`: 
```
openjdk version "1.8.0_252"
OpenJDK Runtime Environment Corretto-8.252.09.1 (build 1.8.0_252-b09)
OpenJDK 64-Bit Server VM Corretto-8.252.09.1 (build 25.252-b09, mixed mode)
```
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.6.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
```
Maven home: .../.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 1.8.0_252, vendor: Amazon.com Inc., runtime: .../java/corretto-8.252.09.1/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.4.0-40-generic", arch: "amd64", family: "unix"
```

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #10812

Supersedes #10970  (this PR adds a commit that pulls the tests from that PR in a separate commit)

### CODE REVIEW

**Comment by loicmathieu** [comment]
As I understand it, your change is broader than #10970 as it also handle differently the creation of the MongoDB client itself.
Can you explain what you did and why in a few sentences please ?

**Comment by geoand** [comment]
Yes of course:

The basic problem that this PR addresses is the handling of the MongoClient instances from Panache. 
To that end it does the following:

* Makes sure that a `MongoClient` beans is created when `clientName` is used on `MongoEntity` (this is similar to the use of `@MongoClientName` in the regular Mongo extension)
* Updates `MongoOperations` to always select the proper MongoClient bean
* Removes the need of reading configuration directly - as the information is already available in CDI in the form of `MongoClientConfig`

The change to use `MongoUnremovableClientsBuildItem` isn't strictly necessary, it's just less code :)

The of course the second commit simply brings in the tests that were added by @ekans as is

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java [hunk 1]
```diff
@@ -129,12 +133,28 @@ ReflectiveHierarchyBuildItem registerForReflection(CombinedIndexBuildItem index)
...
-        unremovable.produce(
-                new UnremovableBeanBuildItem(
-                        new UnremovableBeanBuildItem.BeanClassNamesExclusion(new HashSet<>(
-                                Arrays.asList(MongoClient.class.getName(), ReactiveMongoClient.class.getName())))));
+    void unremoveableClients(BuildProducer<MongoUnremovableClientsBuildItem> unremovable) {
+        unremovable.produce(new MongoUnremovableClientsBuildItem());
+    }
 
+    @BuildStep
+    public void mongoClientNames(ApplicationArchivesBuildItem applicationArchivesBuildItem,
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java:141]: Shouldn't it be a `CombinedIndexBuildItem` to allow defining entities inside a third party library ?

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java [hunk 2]
```diff
@@ -299,36 +301,34 @@ private static ReactiveMongoCollection mongoCollection(Object entity) {
...
-                if (defaultDatabaseName == null) {
-                    defaultDatabaseName = ConfigProvider.getConfig()
-                            .getValue(MONGODB_DATABASE, String.class);
+    private static String getDefaultDatabaseName(MongoEntity entity) {
+        final String beanName = beanName(entity);
+        return defaultDatabaseName.computeIfAbsent(beanName, new Function<String, String>() {
+            @Override
+            public String apply(String s) {
+                MongoClients mongoClients = Arc.container().instance(MongoClients.class).get();
+                MongoClientConfig matchingMongoClientConfig = mongoClients.getMatchingMongoClientConfig(beanName);
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java:318]: This should use the `s` parameter (better rename it key by the way) to avoid a synthetic accessor

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java [hunk 3]
```diff
@@ -299,36 +301,34 @@ private static ReactiveMongoCollection mongoCollection(Object entity) {
...
-            for (Bean<?> bean : beans) {
-                if (bean.getName() != null) {
-                    return (ReactiveMongoClient) Arc.container().instance(entity.clientName()).get();
+                MongoClientConfig defaultMongoClientConfig = mongoClients
+                        .getMatchingMongoClientConfig(MongoClientBeanUtil.DEFAULT_MONGOCLIENT_NAME);
+                if (defaultMongoClientConfig.database.isPresent()) {
+                    return defaultMongoClientConfig.database.get();
                 }
+
+                throw new IllegalArgumentException("The database property was not configured for " + entity);
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java:329]: I think we need to better explain what went wrong as the rules are a little complex. And don't forget that `entity` here can be null so at least you need to provides different error message in case an `entity` configuration exist (so you can add the name of the entity inside the error message) and none are set.

Normally, if we are here, this means that no `database` is configured via the entity annotation so one must be set inside the application.properties, we shoud compute the needed property and log it.

**Review by loicmathieu — CHANGES_REQUESTED** [review]
Overall looks good.

I'm surprised however that we need to ask for the creation of the MongoClient from the Panache extension as I thought this has been fixed a while ago (creating all clients from the properties instead of the injection points).

I request some changes on the `ReactiveMongoOperations` the same apply to the `MongoOperations` but I didn't duplicate them.

**Comment by loicmathieu** [comment]
There is a big PR for Kotlin, @ekans do you need this fix asap or can it wait Quarkus 1.8 (so in one month and a half) ?
If it can wait, @geoand please put this one on-ice untill the Kotlin PR is merged.

**Comment by geoand** [comment]
TBH, I would much prefer this go in before the Kotlin PR since this is a bug fix we want ASAP - as opposed to the Kotlin PR which is a new feature.

Would the Kotlin PR be affected by this a lot? I haven't seen it, but the changes here to the visible parts are pretty minimal (and I can redo them if necessary against whatever branch the PR is in)

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java (hunk 1 above)
Comment (geoand) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java:141]: I pretty much did what we already do for `@MongoClientName`, so I propose we keep it like this for now and change in both places if it comes up in the future.

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java (hunk 3 above)
Comment (geoand) [reply extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java:329]: Makes sense, I'll try and make it clearer

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java (hunk 2 above)
Comment (geoand) [reply extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java:318]: Good catch

File: extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java (hunk 1 above)
Comment (loicmathieu) [reply extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java:141]: OK, if (when) and issue arise we could fix this later

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java (hunk 3 above)
Comment (geoand) [reply extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java]: What do you think about it now?

**Comment by loicmathieu** [comment]
@geoand the Kotlin PR is a very big one and it touches almost everything, you can check it here: #11043. Thanks to propose your help to do the integration work I'm sure @evanchooly will be happy to have your help.

@evanchooly as you wrote the Kotlin PR and this one will implies some changes on the Kotlin PR if it is merged before it I'll follow your advice. Keep in mind that this one could still be integrated inside 1.7 as opposed to the Kotlin one that would target 1.8.

**Comment by geoand** [comment]
Thanks for being vigilant @loicmathieu and seeing the big picture!

As we want to prioritize bug fixes for 1.7, I think we really need this in and if it's needed, I can certainly redo the change in Justin's branch if the a lot of conflicts

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                return defaultMongoClientConfig.database.get();
+            }
+        }
+
+        if (entity == null) {
+            throw new IllegalArgumentException(
+                    "The database property was not configured for the default Mongo Client (via 'quarkus.mongodb.database'");
+        }
+        if (entity.clientName().isEmpty()) {
+            throw new IllegalArgumentException("The database method was not set for : " + entity
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java:86]: Here `MongoEntity` is an annotation instance so I'll rename to something like this

```
The database attribute was not set for the @MongoEntity annotation and neither ...
```

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                    + " and neither was the database property configured for the default Mongo Client (via 'quarkus.mongodb.database')");
+        }
+        throw new IllegalArgumentException(String.format(
+                "The database method was not set for : %s and neither was the database property configured for the named Mongo Client (via 'quarkus.mongodb.%s.database')",
```
Comment (loicmathieu) [inline extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java:90]: Same here I would rephrase to something like

```
The database attribute was not set for the @MongoEntity annotation and neither ...
```

**Comment by ekans** [comment]
> There is a big PR for Kotlin, @ekans do you need this fix asap or can it wait Quarkus 1.8 (so in one month and a half) ?
> If it can wait, @geoand please put this one on-ice untill the Kotlin PR is merged.

@loicmathieu Yes I even want this fix for yesterday!
I will understand if the fix is postponed to 1.8.

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java (hunk 4 above)
Comment (geoand) [reply extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java]: Done

File: extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java (hunk 5 above)
Comment (geoand) [reply extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/BeanUtils.java]: Done

**Review by loicmathieu — APPROVED** [review]
LGTM

**Comment by evanchooly** [comment]
It *looks* like this PR won't be too big of a headache to merge with my PR.  There's the processor and the two operations classes but those look to be easy enough to merge especially since my changes to the operations classes have been mostly confined to unstaticing all the things.  In any event, if my PR is getting bumped to 1.8 (as looks likely) I'll have time to rebase against this change.

**Comment by geoand** [comment]
Yeah, I don't expect this to be a merge headache, as the touch points are pretty small.

In any case, if you need help with it, don't hesitate to ping me.

And just as a reminder, moving to 1.8 just means that your PR can't be merged to `master` until `1.7.0.CR1` is released - so it should only be open for a couple days more.

**Comment by evanchooly** [comment]
Yep.  A bit disappointed that it's getting pushed but that's how it goes sometimes.  A few days won't matter that much anyway.
